### DISCUSSION TITLE
I2S with callback stream

### DISCUSSION THREAD

**Discussion by fobi55** [pschatzmann/arduino-audio-tools#1765] [Q&A] [answered]
Hello,
I have little problem using I2S from codec (Pmod I2S2) with CallbackStream.

_I2SStream codec_in_out;
CallbackStream cb(codec_in_out, IQIF);  // It has output
StreamCopy copier(cb, codec_in_out);_

Should I define separated streams for in/out and how to assign the pins?
With NodeMCU-32S I am using the next pins:

  _auto config = codec_in_out.defaultConfig(RXTX_MODE);
  config.sample_rate = sample_rate;
  config.bits_per_sample = 16;
  config.i2s_format = I2S_STD_FORMAT;
  config.is_master = true;
  config.port_no = 0;
  config.pin_ws = 18;
  config.pin_bck = 5;
  config.pin_data = 19;
  //config.pin_data_rx = 17;
  config.pin_data_rx = 21;
  //config.pin_mck = 0;
  //config.pin_mck = 4;  // blocking com port
  config.pin_mck = 3;    // It works !!!!
  //config.use_apll = true;  
  config.use_apll = false;_

**Comment by pschatzmann** [comment, marked as answer]
In your example you define I2S as in douplex mode: so you can read and write to it!
In your logic however you only read from it: the writing part is missing!

There is an example whre I just [copy the data from i2s to i2s](https://github.com/pschatzmann/arduino-audio-tools/blob/main/examples/examples-stream/streams-i2s-i2s/streams-i2s-i2s.ino)!
